tsl/src/data_node.c

```diff
@@ -2113,3 +2113,33 @@ data_node_name_list_check_acl(List *data_node_names, AclMode mode)
 		}
 	}
 }
+
+/*
+ * Execute a command on a data node.
+ *
+ * Mostly for debugging connection execution functions.
+ */
+Datum
+data_node_exec(PG_FUNCTION_ARGS)
+{
+	const char *nodename = PG_GETARG_CSTRING(0);
+	TSConnection *conn = data_node_get_connection(nodename, REMOTE_TXN_NO_PREP_STMT, true);
+	const text *cmdstr;
+	PGresult *res;
+
+	if (PG_ARGISNULL(1))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("command string cannot be NULL")));
+
+	cmdstr = PG_GETARG_TEXT_P(1);
```

Comment:
This is not safe I feel. What if the cmdstr contains a `DML` command?

Reply:
Not sure why that is an issue. Any command execution is still subject to regular permission control, so a user can only do things it is allowed to.

Already, you can use `distributed_exec` to execute any statement on data nodes.